src/Raven.Server/Documents/Handlers/AttachmentHandler.cs

```diff
@@ -0,0 +1,207 @@
+                {
+                    HttpContext.Response.StatusCode = (int)HttpStatusCode.NotModified;
+                    return;
+                }
+
+                HttpContext.Response.Headers[Constants.Headers.Etag] = "\"" + attachment.Etag + "\"";
+
+                // Ensure that files are not cached at the browser side.
+                HttpContext.Response.Headers["Cache-Control"] = "no-cache, no-store, must-revalidate";
+                HttpContext.Response.Headers["Expires"] = "0";
+
+                HttpContext.Response.Headers["Content-Type"] = attachment.ContentType.ToString();
+
+                var stream = Database.DocumentsStorage.GetAttachmentStream(context, attachment.StreamIdentifier);
+                if (stream == null)
+                {
+                    throw new FileNotFoundException("Attachment is not found. This should not happen.");
+                }
+
+                var fileName = Path.GetFileName(attachment.Name);
+                HttpContext.Response.Headers["Content-Disposition"] = $"attachment; filename=\"{fileName}\"";
+                // TODO: HttpContext.Response.Headers["Content-Range"] = ;
+
+                await stream.CopyToAsync(ResponseBodyStream());
+            }
+        }
+
+        [RavenAction("/databases/*/attachments", "PUT")]
+        public async Task Put()
+        {
+            DocumentsOperationContext context;
+            using (ContextPool.AllocateOperationContext(out context))
+            {
+                var id = GetQueryStringValueAndAssertIfSingleAndNotEmpty("id");
+                var name = GetQueryStringValueAndAssertIfSingleAndNotEmpty("name");
+                var contentType = GetStringQueryString("contentType", false) ?? "";
+
+                var tempPath = GetUniqueTempFileName(Path.GetTempPath(), "tmp");
+                using (var file = new FileStream(tempPath, FileMode.CreateNew, FileAccess.ReadWrite, FileShare.None, 4096, FileOptions.DeleteOnClose) )
+                {
+                    await RequestBodyStream().CopyToAsync(file, 81920, Database.DatabaseShutdown);
+                    file.Position = 0;
+
+                    var etag = GetLongFromHeaders("If-Match");
+
+                    var cmd = new MergedPutAttachmentCommand
+                    {
+                        Database = Database,
+                        ExpectedEtag = etag,
+                        DocumentId = id,
+                        Name = name,
+                        Stream = file,
+                        ContentType = contentType,
+                    };
+                    await Database.TxMerger.Enqueue(cmd);
+                    cmd.ExceptionDispatchInfo?.Throw();
+
+                    HttpContext.Response.StatusCode = (int)HttpStatusCode.Created;
+
+                    using (var writer = new BlittableJsonTextWriter(context, ResponseBodyStream()))
+                    {
+                        writer.WriteStartObject();
+
```

Comment:
I'm thinking about the debuggability of this over the wire, so that would make things easier. Also make sure that we get the proper casing.